File: l3.ml
(*zadanie 1*)
let rec splitBySign list =
  match list with
    [] -> ([],[])
  | head:: tail -> let (negative, positiveOdd) = splitBySign tail in
                   if head < 0 then (head::negative, positiveOdd)
                   else if head mod 2 = 1 then (negative, head::positiveOdd)
                   else (negative, positiveOdd);;

splitBySign [-3;-6;7;-9;13] = ([-3;-6;-9],[7;13]);;
splitBySign [] = ([],[]);;
splitBySign [1;2;-3] = ([-3],[1]);;
splitBySign [1;3] = ([],[1;3]);;

(*zadanie 2*)
let rec lengthOfList list =
  if list = [] then 0
  else 1 + lengthOfList(List.tl list);;

lengthOfList[] = 0;;
lengthOfList[1;2] = 2;;
lengthOfList["a";"b";"c"] = 3;;
lengthOfList[5;4;3;2] = 4;;

(*zadanie 3*)
let rec joinList list1 list2 =
  match(list1,list2) with
    ([],[]) -> []
  | ([],_) -> list2
  | (_,[]) -> list1
  | (head1::tail1, head2::tail2) -> head1::head2::joinList(tail1 tail2);;

joinList [5;4;3;2] [1;2;3;4;5;6] = [5; 1; 4; 2; 3; 3; 2; 4; 5; 6];;
joinList [] ["a";"b"] = ["a";"b"];;
joinList [1;2;3] [] = [1;2;3];;
joinList [1;3;5;7] [2;4] = [1; 2; 3; 4; 5; 7];;

File: l3.scala
object l3 {
  //1
  //O(n^2)
  //Memory O(1)
  def splitBySign[A](xs: List[Int])={
    def split(xs: List[Int], negativeList:List[Int], positiveList:List[Int]):(List[Int], List[Int])={
      if xs ==Nil then (negativeList, positiveList)
      else if xs.head<0 then split(xs.tail, negativeList:::List(xs.head), positiveList)
      else if xs.head>0 && xs.head %2==1 then split(xs.tail, negativeList, positiveList:::List(xs.head))
      else split(xs.tail, negativeList,  positiveList)
    }
    split(xs, List(), List())
  }

  //2
  //O(n)
  //Memory O(1)
  def lengthOfList[A](xs: List[A]): Int={
    if xs!=Nil then 1+lengthOfList(xs.tail)
    else 0
  }

  //3
  //O(n)
  //Memory O(n)
  def joinList[A](xs: List[A], ys: List[A]): List[A]={
    if xs!=Nil && ys!=Nil then xs.head::ys.head::joinList(xs.tail, ys.tail)
    else if xs!=Nil && ys==Nil then xs
    else if xs==Nil && ys!=Nil then ys
    else Nil
  }

  def l3(args: Array[String]): Unit = {
    println(splitBySign(List(1,2,-3, 10,-7,9,10,-11, 13)))
    println(splitBySign(List()))
    println(splitBySign(List(1,1,-1,-1,1,10,-11,11,12,14)))
    println(" ")
    println(lengthOfList(List(5,4,3,2)))
    println(lengthOfList(List()))
    println(lengthOfList(List(1,2,3,4,5,6,7,8,9)))
    println(" ")
    println(joinList(List(5, 4, 3, 2), List(1, 2, 3, 4, 5, 6)))
    println(joinList(List(1,2,3,4,5,6,7), List(0,11)))
    println(joinList(List(), List()))
    println(" ")
  }
}
